### PR TITLE
Improve deploying from a packrat project

### PR DESCRIPTION
**Pull request by slopp**
This PR aims to improve two use cases:

1. When rsconnect is deploying from a packrat project.
The existing packrat lock file is respected (instead of generating a new one). 

2. When using the RSC API to deploy content pulled from a remote repo
One use case for the RSC API is for Jenkins (or another CI tool) to deploy content to Connect from a remote repository. In this case, rsconnect is not being called from the development environment. The packrat lock file describing the original environment needs to be respected. This packrat lock file can come from a user developing in a packrat project, or from a user running: `packrat::.snapshotImpl(snapshot.sources=FALSE)` prior to committing to the repo.

This PR also checks to see if a .Rprofile is being added that contains `source("packrat/init.R")` and comments out that line if it exists. This is useful because a) packrat shouldn't be initiated on the target server and b) packrat is always ignored by default when adding files to the bundle - this means the default behavior when deploying from a packrat project is to include the .Rprofile but exclude packrat which causes an error

### CODE REVIEW

**Comment by jjallaire** [comment]
I'm just about to submit rsconnect to CRAN. @jmcphers and @kevinushey Could you review this and let me know whether we should take this before or after the CRAN submission? (also of course let us know if there are changes/concerns on the PR itself).

**Comment by jmcphers** [comment]
Without this change, publishing fails on the Connect side if the directory being published is the top level of a Packrat project. There's a customer support ticket on this very issue (not sure if that was the trigger for @slopp's change or merely serendipitous timing). 

For that reason I think we need to take this change in some form prior to CRAN release; if taking the whole thing feels too risky we could take just the minimal portion which works around the failure.

**Comment by kevinushey** [comment]
`rsconnect::performPackratSnapshot()` does a bunch of other things that might be required in this new code path. Some thoughts:

1. We ensure that the `snapshot.recommended.packages` Packrat option is turned on -- this ensures that recommended packages (e.g. `MASS`) enter the lockfile. If someone is working within a packrat project with a pre-existing lockfile, then deployment will likely fail if they're using recommended CRAN packages without snapshotting them. In hindsight it was a mistake in Packrat to control whether certain packages are snapshotted -- we should just snapshot _all_ packages discovered in the active libraries, and let `restore()` be smart about handling base / recommended packages -- but unfortunately that's not how Packrat works today.

2. We eagerly load the `BiocInstaller` package to work around an issue where the package fails to detect `https` support.

File: R/bundle.R [hunk 1]
```diff
@@ -45,6 +56,12 @@ maxDirectoryList <- function(dir, parent, totalSize) {
     contents <- contents[!grepl(glob2rx(".Rhistory"), contents)]
   }
 
+
+  # add packrat lock file to contents if it exists
+  # TODO: Handle the case where dir is a subdirectory of the project
+  if (file.exists(snapshotLockFile(dir)))
```
Comment (jmcphers) [inline R/bundle.R]: Put this in the `nchar(parent) == 0` clause above so that it only runs at the root; in the case of deploying a directory which contains a packrat project in a subdirectory, we already include the lockfile (as the `packrat` folder is only excluded at the root)

File: R/bundle.R [hunk 2]
```diff
@@ -15,9 +15,20 @@ bundleAppDir <- function(appDir, appFiles, appPrimaryDoc = NULL) {
...
+
     if (!file.exists(dirname(to)))
       dir.create(dirname(to), recursive = TRUE)
     file.copy(from, to)
+
+    #ensure .Rprofile doesn't call packrat/init.R
+    if (basename(to) == ".Rprofile") {
+      origRprofile <- readLines(to)
+      newRprofile <- gsub( 'source(\"packrat/init.R\")',
+                           '#source(\"packrat/init.R\")',
```
Comment (jmcphers) [inline R/bundle.R]: Could you add a comment to the `.Rprofile` here indicating why the `source()` line was commented out, so that it's clear what's going on if someone opens the bundle? e.g.

    # Modified by rsconnect package $VERSION on $DATE:
    # Packrat initialization disabled in published application
    # source ...

File: R/bundle.R [hunk 3]
```diff
@@ -485,7 +502,11 @@ addPackratSnapshot <- function(bundleDir, implicit_dependencies = c()) {
 
   # generate the packrat snapshot
   tryCatch({
-    performPackratSnapshot(bundleDir)
+    if (!file.exists(snapshotLockFile(bundleDir))) {
+      performPackratSnapshot(bundleDir)
+    } else {
+      message("Re-using Packrat Lock File")
```
Comment (jmcphers) [inline R/bundle.R]: Do we need to check that the lockfile is up to date before using it? i.e. what if my packrat status isn't clean (I'm using packages I haven't yet snapshotted) and I try to publish?

**Comment by slopp** [comment]
My original thought was the user would be guaranteeing the state of the packrat lock file. ie, In the remote deployment use case (where the system calling rsconnect::deployApp is not the environment where the code was created), the user would've  run performPackratSnapshot and committed the result.

FWIW - It seems a bit awkward that performPackratSnapshot lives in rsconnect and is so different from packrat::snapshot.  Would the best practice be to always do what performPackratSnapshot is doing?

The bigger idea:  I think it is convenient to use packrat the way rsconnect is using it - to generate a manifest - as opposed to the full blown implementation of packrat. For example, I can imagine a world where every time I commit R code for an analysis I also commit the result of running performPackratSnapshot, but I don't actually use packrat mode. In this world, I would've already run performPackratSnapshot before calling rsconnect::deployApp. 

This PR implements the necessary de-coupling of deployApp and performPackratSnapshot. Perhaps it would make more sense to surface this as an option, ie packrat::deployApp(..., exisiting.packrat.lock = "/path/to/packrat.lock") so the behavior is intentionally invoked instead of being the default? 

All that said, the only specific fix necessary today is in: https://github.com/rstudio/rsconnect/pull/143/commits/aee35b76e053fb6587a8e2605c5c47a791cd254b

**Comment by kevinushey** [comment]
> My original thought was the user would be guaranteeing the state of the packrat lock file. ie, In the remote deployment use case (where the system calling rsconnect::deployApp is not the environment where the code was created), the user would've run performPackratSnapshot and committed the result.

This makes sense if we want to push the onus of generating a correct lockfile on to the user, but I'm just worried that there may be some traps going this route -- ie, weird deployment errors that ultimately put extra strain on the support team. (It may be fine to just tackle them as they come up)

> FWIW - It seems a bit awkward that performPackratSnapshot lives in rsconnect and is so different from packrat::snapshot. Would the best practice be to always do what performPackratSnapshot is doing?

Perhaps `rsconnect` needs something like a `prepareForPackratSnapshot()` that does the steps outlined before -- does that make sense? Or should `rsconnect` just export its own `performSnapshot()` function?

In any case, `rsconnect` has its own snapshot function since the type of snapshot you want for a deployment may differ from the type of snapshot you want for other types of project management.

> The bigger idea: I think it is convenient to use packrat the way rsconnect is using it - to generate a manifest - as opposed to the full blown implementation of packrat. For example, I can imagine a world where every time I commit R code for an analysis I also commit the result of running performPackratSnapshot, but I don't actually use packrat mode. In this world, I would've already run performPackratSnapshot before calling rsconnect::deployApp.

One of the things that makes this difficult is that Packrat adopts a per-directory project view, whereas `rsconnect` deployments can be single files (perhaps sharing the same folder). The package requirements for each file might differ. That said, having a lockfile that captures the union of required dependencies in a folder might be fine?

> This PR implements the necessary de-coupling of deployApp and performPackratSnapshot. Perhaps it would make more sense to surface this as an option, ie packrat::deployApp(..., exisiting.packrat.lock = "/path/to/packrat.lock") so the behavior is intentionally invoked instead of being the default?

This seems reasonable, albeit with the caveats of right now us effectively telling the user "if it breaks, you get to keep the pieces".

> All that said, the only specific fix necessary today is in: aee35b7

---

tl;dr: I think this PR is okay, I'm just worried about the traps that users adopting this workflow could encounter.

File: R/bundle.R (hunk 3 above)
Comment (slopp) [inline R/bundle.R]: I think this is a pretty easy addition `capture.output(packrat::status())` would do it. But what should the default behavior be - try to use the lockfile and flag a warning, or fall back on performPackratSnapshot?

**Comment by slopp** [comment]
Turns out, this implementation does not address the two use cases I was hoping for. This is because starting with an existing lock file is not enough for deployment. (rsconnect also assumes the packages are installed to compile the description files and dependencies). For this reason, the ability to provide a lock file is not sufficient and introduces some potential problems if you were to deploy from a packrat repository when packrat mode is off.  I have reverted some changes, only the bug fix needs to be merged.

TL;DR:  Reverted the majority of changes, only leaving the bug fix that checks .Rprofile for packrat/init.R

**Comment by jjallaire** [comment]
@jmcphers and @kevinushey Are we okay to merge this now with the more narrowly scoped fix? (will be submitting to CRAN early next week).

**Comment by jmcphers** [comment]
Yes, I'm a fan of taking the narrower fix. I think we should still entertain the idea of using the local packrat lockfile and library to drive publish as it enables a few publishing scenarios that are just not possible today, but it now seems clear that there are enough subtleties and edge cases that we'd want to take our time with that change.
